### DISCUSSION TITLE
CSV map error.

### DISCUSSION THREAD

**Discussion by erikhman** [HaxeFlixel/flixel#2439] [Q&A] [answered]
I recently started programming in haxeflixel, and, like most of those who started doing this recently, the reason was FNF modding. But the knowledge that can be found in groups dedicated to it, are limited to problems typical for FNF. Therefore, it is difficult to find solutions there for the problems that arise when implementing original ideas. And that's exactly what I'm facing. I want to add a simple platformerState into the game to diversify the gameplay. But I'm having problems with the map. When I try to load a CSV file with a map, the game, for some reason, cannot find it.
Here i am trying to load the map in.
```
class WalkingState extends FlxState
{  
    var levelBounds:FlxGroup;
    var player:Player;
    var map:FlxTilemapExt;

    public function new(levelName:String)
    {
        super();
        switch (levelName)
        {
            case 'test':
            { 
                bgColor = 0xffcccccc;
                player = new Player(200, 200);
                map = new FlxTilemapExt();
                add(player);
                add(map.loadMapFromCSV("shared/images/map.csv", "shared/images/1.png", 16, 16));
                trace("Loaded succsesfully");
            }
            default:
            {
               bgColor = 0xffcccccc;
                player = new Player(200, 200);
                map = new FlxTilemapExt();
                add(player);
                add(map.loadMapFromCSV("shared/images/map.csv", "shared/images/1.png", 16, 16));
                trace("Loaded succsesfully");
            }
        }
    }
.
.
.
}
```
And here is the part of the code, where the error occurs.

```
public function loadMapFromCSV(MapData:String, TileGraphic:FlxTilemapGraphicAsset, TileWidth:Int = 0, TileHeight:Int = 0, ?AutoTile:FlxTilemapAutoTiling,
			StartingIndex:Int = 0, DrawIndex:Int = 1, CollideIndex:Int = 1)
	{
		// path to map data file?
		
		if (Assets.exists(MapData))    //<----For some reason it returns false
		{
			MapData = Assets.getText(MapData);
		}
.
.
.
	}
```

Moreover, everything works in this HaxeFlixel Demo (https://haxeflixel.com/demos/FlxTilemapExt/). As far as I understand, the problem should be in the project file, but I do not know what exactly i need to change in there. Here is a link to the vanilla version of the game I am modifying https://github.com/KadeDev/Kade-Engine/releases/tag/1.7. And, in case my theory with project file is correct, here is project file(as a txt, because i cant attach xml).
[project.txt](https://github.com/HaxeFlixel/flixel/files/7478345/project.txt)

**Comment by RichardBray** [comment]
I haven't created a level from csv before but looking at this codebase
https://github.com/HaxeFlixel/flixel-demos/blob/master/Platformers/FlxTilemapExt/source/PlayState.hx#L47

It looks like the file has been changed from .csv to .txt. Is this something you have tried?

**Reply by RichardBray** [reply]
Also, it would be good to check if you're asset path is correct. The reason the link I shared in the answer works is because the asset path has been specified in the Project.xml file
https://github.com/HaxeFlixel/flixel-demos/blob/master/Platformers/FlxTilemapExt/Project.xml#L34

I don't know if your `shared` path has been added there as well.

**Reply by erikhman** [reply]
I already tried changing it to .txt and it didnt work.
About asset path. I attached the Project.xml file as a txt one to the message above. I'm almost 100% sure, that the error is somehow related to the file. But you could see that the `shared` path is specified in there, so, i guess, there is something else that i'm missing.

**Reply by erikhman** [reply]
Also, in the HaxeFlixel demo you mentioned, I noticed some .hxproj file. 
https://github.com/HaxeFlixel/flixel-demos/blob/master/Platformers/FlxTilemapExt/FlxTilemapExt.hxproj
Maybe that's the problem? I mean, I don't have one in my build.

**Reply by RichardBray** [reply]
Interesting spot. I'm not sure the .hxproj file has anything to do with it but it's worth asking @Gama11 who maintains the demos.

**Reply by Gama11** [reply]
.hxproj is not relevant, it's a FlashDevelop / HaxeDevelop project file (and has no impact on compilation).

**Comment by Geokureli** [comment, marked as answer]
can you tell us more about your assets folder structure? I noticed `shared/images/` which is similar to the FNF assets layout, which is a very non-standard way of doing it. also what does your project.xml assets path look like?

If you're on VSCode you can right click any file in the left side file tree and get the relative path. I use this a lot because it's easy to misspell a file or type in the wrong path. if that doesn't work try taking a screenshot of the file in the file tree and posting that here

**Reply by erikhman** [reply]
Here is my assets path from project.hx:

```
	<!-- _____________________________ Path Settings ____________________________ -->

	<set name="BUILD_DIR" value="export/debug" if="debug" />
	<set name="BUILD_DIR" value="export/release" unless="debug" />
	<set name="BUILD_DIR" value="export/32bit" if="32bit" />

	<classpath name="source" />

	<assets path="assets/preload" rename="assets" exclude="*.ogg" if="web"/>
	<assets path="assets/preload" rename="assets" exclude="*.mp3" unless="web"/>
	
	<!-- <define name="PRELOAD_ALL" /> -->
	<define name="PRELOAD_ALL" unless="web" />
	<define name="NO_PRELOAD_ALL" unless="PRELOAD_ALL"/>
	
	<section if="PRELOAD_ALL">
		<library name="songs"    preload="true" />
		<library name="shared"   preload="true" />
		<library name="tutorial" preload="true" />
		<library name="week1"    preload="true" />
		<library name="week2"    preload="true" />
		<library name="week3"    preload="true" />
		<library name="week4"    preload="true" />
		<library name="week5"    preload="true" />
		<library name="week6"    preload="true" />
		<library name="sm"    preload="true" />
	</section>
	
	<section if="NO_PRELOAD_ALL">
		<library name="songs"    preload="false" />
		<library name="shared"   preload="false" />
		<library name="tutorial" preload="false" />
		<library name="week1"    preload="false" />
		<library name="week2"    preload="false" />
		<library name="week3"    preload="false" />
		<library name="week4"    preload="false" />
		<library name="week5"    preload="false" />
		<library name="week6"    preload="false" />
		<library name="sm"    preload="false" />
	</section>
	
	<assets path="assets/songs"    library="songs"    exclude="*.ogg" if="web"/>
	<assets path="assets/songs"    library="songs"    exclude="*.mp3" unless="web"/>
	<assets path="assets/shared"   library="shared"   exclude="*.ogg" if="web"/>
	<assets path="assets/shared"   library="shared"   exclude="*.mp3" unless="web"/>
	<assets path="assets/tutorial" library="tutorial" exclude="*.ogg" if="web"/>
	<assets path="assets/tutorial" library="tutorial" exclude="*.mp3" unless="web"/>
	<assets path="assets/week1"    library="week1"    exclude="*.ogg" if="web"/>
	<assets path="assets/week1"    library="week1"    exclude="*.mp3" unless="web"/>
	<assets path="assets/week2"    library="week2"    exclude="*.ogg" if="web"/>
	<assets path="assets/week2"    library="week2"    exclude="*.mp3" unless="web"/>
	<assets path="assets/week3"    library="week3"    exclude="*.ogg" if="web"/>
	<assets path="assets/week3"    library="week3"    exclude="*.mp3" unless="web"/>
	<assets path="assets/week4"    library="week4"    exclude="*.ogg" if="web"/>
	<assets path="assets/week4"    library="week4"    exclude="*.mp3" unless="web"/>
	<assets path="assets/week5"    library="week5"    exclude="*.ogg" if="web"/>
	<assets path="assets/week5"    library="week5"    exclude="*.mp3" unless="web"/>
	<assets path="assets/week6"    library="week6"    exclude="*.ogg" if="web"/>
	<assets path="assets/week6"    library="week6"    exclude="*.mp3" unless="web"/>
	<assets path="assets/sm"    library="sm"    exclude="*.ogg" if="web"/>
	<assets path="assets/sm"    library="sm"    exclude="*.mp3" unless="web"/>

	<assets path='example_mods' rename='mods' embed='false'/>
	<assets path='art/readme.txt' rename='do NOT readme.txt' />
	<assets path='LICENSE' rename='LICENSE.txt' />
	<!-- <template path='mods' /> -->
	
	<!-- <assets path="CHANGELOG.md" rename='changelog.txt'/> -->

	<!-- NOTE FOR FUTURE SELF SINCE FONTS ARE ALWAYS FUCKY 
		TO FIX ONE OF THEM, I CONVERTED IT TO OTF. DUNNO IF YOU NEED TO
		THEN UHHH I USED THE NAME OF THE FONT WITH SETFORMAT() ON THE TEXT!!!
		NOT USING A DIRECT THING TO THE ASSET!!!
	-->
	<assets path="assets/fonts" embed='true'/>
```
And here is my file tree:
![image](https://user-images.githubusercontent.com/78647459/140554272-797ee15e-6070-47f8-af65-34e90d94ff95.png)

**Reply by Geokureli** [reply]
So this is a heavy FNF mod. sorry, I didn't fully read your original post. FNF does a lot of asset management in order to keep the initial web preload down to a minimum. shared gets loaded whenever you load ANY level, and contains assets that appear in multiple levels.

> If you're on VSCode you can right click any file in the left side file tree and get the relative path. 

When you right click and get the relative path you would have seen the path is actually `assets/shared/images/map.csv`, but because FNF has different libraries that load at different times it gets a little hairy. **I strongly recommend making a platformer in a brand new project with the default project setup and assets structure, first. I can tell you're new to this so this might be an easier way to learn**. once you have a handle on that (or if you still want to work on the FNF mod directly), learn more about OpenFLs library/asset system. Note that you can put everything in the preload folder and access it via the `Paths` class. Which is a Funkin class that manages all the different libraries.

```hx
add(map.loadMapFromCSV(Paths.files("images/map.csv"), Paths.image("images/1.png"), 16, 16));
```

**Reply by erikhman** [reply]
Thanks man, it worked! Ive been thinking about using Paths class, but i failed to find the right function for .csv files, so i gave up on that idea.

But you're right, I'm new to this, it would be better for me to create a platformer part separately from FNF and then just import it.

**Reply by Geokureli** [reply]
Plus, platformers are way more fun than mods

**Reply by erikhman** [reply]
May i ask you one final question? As you could see, i am using FlxTilemapExt, because of slopes collision support. Originally i was planning to implement curves into the game, like, make the ground a smooth sin(x)-like line. But i didnt find any examples of such idea in other games, so i moved on from curves to slopes. But i am not that experienced in Flixel, and in game developing in general. Is there some way to add curves into the game?

**Reply by Geokureli** [reply]
Ive never used tilemapExt, you might be better off asking in a new question

**Reply by RichardBray** [reply]
@erikhman For slopes, I wouldn't recommend using FlxTilemapExt since it's quite limited. I'd say use something like [Echo](https://github.com/AustinEast/echo-flixel) which is an actual physics library.

**Reply by erikhman** [reply]
@RichardBray It actually looks much better than "flxtilemapext". I'll try to work with it, but just in case, I'd like to ask something.
Initially, I wanted to implement something like what is shown in the picture: the character should be able to move along a curved surface, while its sprite should rotate, depending on the derivative at current point of the curve.  Do you happen to know how to implement it? Obviously, there is mathematics involved, so it seems to me that there must surely be an engine or library that solves this problem...
<img width="1879" alt="example" src="https://user-images.githubusercontent.com/78647459/140758103-55b87cfc-eca5-411c-bb18-6ae14b8e926c.png">

**Reply by RichardBray** [reply]
Ha, maybe this is a question for a new discussion thread but I think Echo can solve those problems. I haven't tried it with sprite rotation on slopes but I know that .

Even just following the instructions on this page https://gist.github.com/AustinEast/524db026a4fea298a0eebf19459131cc can produce some pretty decent slope physics.

https://user-images.githubusercontent.com/1377253/140799820-cc356632-c861-4ce7-83e2-8c35b95194fa.mp4
